File: mobile_app/lib/shared/models/lyrics.dart
class SyncedLyricLine {
  final int timeMs;
  final String text;

  const SyncedLyricLine({required this.timeMs, required this.text});

  factory SyncedLyricLine.fromJson(Map<String, dynamic> json) {
    return SyncedLyricLine(
      timeMs: (json['time_ms'] as num?)?.toInt() ?? 0,
      text: json['text']?.toString() ?? '',
    );
  }
}

/// Mirrors app/schemas/lyrics.py LyricsResponse.
class LyricsData {
  final String songId;
  final bool hasLyrics;
  final bool isSynced;
  final String? plainText;
  final List<SyncedLyricLine> syncedLines;
  final String? language;

  const LyricsData({
    required this.songId,
    required this.hasLyrics,
    required this.isSynced,
    this.plainText,
    this.syncedLines = const [],
    this.language,
  });

  factory LyricsData.fromJson(Map<String, dynamic> json) {
    return LyricsData(
      songId: json['song_id']?.toString() ?? '',
      hasLyrics: json['has_lyrics'] == true,
      isSynced: json['is_synced'] == true,
      plainText: json['plain_text']?.toString(),
      syncedLines: (json['synced_lines'] as List?)
              ?.map((e) => SyncedLyricLine.fromJson((e as Map).cast<String, dynamic>()))
              .toList() ??
          const [],
      language: json['language']?.toString(),
    );
  }
}

File: app/schemas/lyrics.py
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


class SyncedLine(BaseModel):
    time_ms: int = Field(..., ge=0)
    text: str


class LyricsUpsertRequest(BaseModel):
    plain_text: Optional[str] = None
    synced_lines: Optional[List[SyncedLine]] = None
    language: Optional[str] = Field(None, max_length=50)
    source: str = Field("manual", max_length=100)


class LyricsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    song_id: str
    has_lyrics: bool
    is_synced: bool
    plain_text: Optional[str] = None
    synced_lines: Optional[List[SyncedLine]] = None
    language: Optional[str] = None
    source: Optional[str] = None
